File: prototype/bin/Prototype.js
class Shape {
    constructor(source) {
        this.source = source;
    }

    clone() {
        return new Shape(this.source);
    }
}

class Rectangle extends Shape {
    constructor(width, height) {
        super({ width, height });
    }
}

class Circle extends Shape {
    constructor(radius) {
        super({ radius });
    }
}

const shapes = [
    new Rectangle(100, 200),
    new Circle(50),
    new Circle(30),
    new Rectangle(60, 30),
];

const shapesCopy = shapes.map(shape => shape.clone());

console.dir({ shapes });

console.dir({ shapesCopy });

if (shapes[0] === shapesCopy[0]) {
    console.log('Shapes are the same');
} else {
    console.log('Shapes are not the same');
}

File: prototype/src/Prototype.py
import copy

class Shape:
    def __init__(self):
        self.type = None

    def clone(self):
        return copy.copy(self)
    
    def getType(self):
        return self.type
    
    def draw(self):
        print("Drawing a shape")
        
class Rectangle(Shape):
    def __init__(self):
        self.type = "Rectangle"
        
    def draw(self):
        print("Drawing a rectangle")
        
        
class Circle(Shape):
    def __init__(self):
        self.type = "Circle"
        
    def draw(self):
        print("Drawing a circle")
        
        

original = Rectangle()
print(original.getType())
original.draw()

clone = original.clone()
print(clone.getType())
clone.draw()

if original is clone:
    print("Original and clone are the same object")
else:
    print("Original and clone are different objects")